chruby.1
.TH "CHRUBY" "1" "January 2013" "postmodern" "Ruby Programmers Reference Guide"
.BR chruby (1) 
\-\- Changes the current Ruby.
.SH SYNOPSIS
.PP
\fB\fCchruby\fR [RUBY|VERSION|system] [RUBY_OPTS]
.SH DESCRIPTION
.PP
Changes the current Ruby version by correctly setting the appropriate environment variables. Additional documentation available at https://github.com/postmodern/chruby/blob/master/README.md
.UR https://github.com/postmodern/chruby/blob/master/README.md
.UE .
.SH OPTIONS
.PP
\fB\fC-h\fR, \fB\fC--help\fR
.SH EXAMPLES
.PP
List available Rubies:
    $ chruby
     ruby\-1.9.3\-p327
     jruby\-1.7.0
     rubinius\-2.0.0\-rc1
.PP
Select a Ruby:
    $ chruby 1.9.3
    $ chruby
     * ruby\-1.9.3\-p327
       jruby\-1.7.0
       rubinius\-2.0.0\-rc1
.PP
Switch to JRuby in 1.9 mode:
    $ chruby jruby \-\-1.9
.PP
Switch back to system Ruby:
    $ chruby system
.PP
Switch to an arbitrary Ruby on the fly:
    $ chruby_use /path/to/ruby
.SH FILES
.PP
~/.gem/$ruby/$version
.PP
/path/to/$ruby/$gemdir
.SH ENVIRONMENT
.PP
PATH Updated to include Rubies and RubyGems bin/ directories.
.PP
GEM_HOME
.PP
GEM_PATH
.PP
RUBY_ROOT
.PP
RUBY_ENGINE
.PP
RUBY_VERSION
.PP
GEM_ROOT
.PP
RUBYOPT Optionally set if second arguement is given.
.SH AUTHORS
.PP
Designed and implemented by Postmodern postmodern\.mod3@gmail\.com
.ME .
.SH SEE ALSO
.PP
.BR ruby (1), 
.BR gem (1)

ChangeLog.md
### 0.3.1 / 2012-12-29

* Fixed the auto-detection of `~/.rubies/*`.
* Check if `bin/ruby` exists and is executable before switching to a Ruby.
* Prevent `export=""` from accidentally being set under [zsh].
* Prevent `script/setup.sh` from exiting if a `brew install` fails because all
  packages are already installed.
* Updated the example `/etc/profile.d/chruby.sh` to only load under [bash]
  and [zsh].

### 0.3.0 / 2012-12-20

* Added the `chruby-exec` utility for use in `crontab` or with Continuous
  Integration (CI).
* Added support for auto-detecting Rubies installed into `/opt/rubies/` or
  `~/.rubies/`.
* Added `share/chruby/auto.sh`, which provides support for auto-switching
  to Rubies specified in the [.ruby-version](https://gist.github.com/1912050)
  file.
* Removed the "short circuit" check in `chruby_use`, to allow forcibly
  switching to the current Ruby, in case `PATH` or `GEM_PATH` become corrupted.

### 0.2.6 / 2012-12-18

* Forcibly switch to system Ruby when loading `share/chruby/chruby.sh`.
  This fixes switching issues for [tmux] users.

### 0.2.5 / 2012-12-15

* Renamed the `RUBY` environment variable to `RUBY_ROOT` to avoid breaking
  the `FileUtils#ruby` method in [rake](http://rake.rubyforge.org/).
* Do not unset `GEM_HOME`, `GEM_PATH`, `GEM_ROOT` if running under root.

### 0.2.4 / 2012-12-13

* Added a `Vagrantfile` for testing chruby in various environments.
* Changed all code and examples to reference `/opt/rubies/`.
* Ensure all error messages are printed to stderr.
* Refactored `scripts/setup.sh` to manually install all Rubies and install any
  dependencies via the System Package Manager.
* PGP signatures are now stored in `pkg/`.

#### Makefile

* Updated the `Makefile` to be compatible with BSD automake.
* Do not override `PREFIX`.
* Added a `test` task.

#### Homebrew

* Use `HOMEBREW_PREFIX`.
* Use `sha1` instead of `md5` (deprecated).
* No longer dynamically generate the example configuration.

### 0.2.3 / 2012-11-19

* Updated the `Makefile` to be compatible with the [dash] shell.
* Use inline substring substitutions instead of `sed`.

### 0.2.2 / 2012-11-17

* Use `typeset` to declare `RUBIES` as an indexed Array.
* Use the correct globbed Array syntax for both [zsh] and [bash].
* Improved the post-installation message in the [homebrew] recipe to auto-detect
  [RVM], [rbenv] and [rbfu].

### 0.2.1 / 2012-10-23

* Fixed `make install` to work on OS X.
* Added a [homebrew] recipe.

### 0.2.0 / 2012-10-16

* Install `chruby.sh` into `$PREFIX/share/chruby/`.

### 0.1.2 / 2012-08-29

* Check if `$RUBY` _and_ `$RUBYOPT` are different from the arguments passed to
  `chruby_use`.
* Fixed a spelling error in the README (thanks Ian Barnett).

### 0.1.1 / 2012-08-24

* Added unit-tests using [shunit2](http://code.google.com/p/shunit2/)
* Improved sanitation of `$PATH` in `chruby_reset`. (thanks mpapis)
* If the desired Ruby is already in use, immediately return from `chruby_use`.
* Export `$RUBY_ENGINE`, `$RUBY_VERSION`, `$GEM_ROOT` in `chruby_use`.

### 0.1.0 / 2012-08-18

* Added support for [zsh].
* Renamed the `$RUBY_PATH` variable to `$RUBY`.
* Set the `$RUBY_ENGINE` variable.
* Set the `$GEM_ROOT` variable to `Gem.default_dir`.
  This supports the custom RubyGems directory used by [Rubinius].
* Only initialize the `$RUBIES` variable if it does not have a value.

### 0.0.2 / 2012-08-14

* Added a `LICENSE.txt`.
* Added a `ChangeLog.md`.
* Updated the `Makefile` to generate proper tar archives.

### 0.0.1 / 2012-08-01

* Initial release.

[dash]: http://gondor.apana.org.au/~herbert/dash/
[bash]: http://www.gnu.org/software/bash/
[zsh]: http://www.zsh.org/
[tmux]: http://tmux.sourceforge.net/

[Rubinius]: http://rubini.us/
[homebrew]: http://mxcl.github.com/homebrew/

[RVM]: https://rvm.io/
[rbenv]: https://github.com/sstephenson/rbenv#readme
[rbfu]: https://github.com/hmans/rbfu#readme


share/chruby/chruby.sh
typeset -a RUBIES

[[ -d /opt/rubies/    ]] && RUBIES+=(/opt/rubies/*)
[[ -d "$HOME/.rubies" ]] && RUBIES+=("$HOME"/.rubies/*)

function chruby_reset()
{
	[[ -z "$RUBY_ROOT" ]] && return

	export PATH=":$PATH:"; export PATH=${PATH//:$RUBY_ROOT\/bin:/:}

	if [[ ! $UID -eq 0 ]]; then
		export PATH=${PATH//:$GEM_HOME\/bin:/:}
		export PATH=${PATH//:$GEM_ROOT\/bin:/:}
		unset GEM_ROOT GEM_HOME GEM_PATH
	fi

	export PATH=${PATH#:}; export PATH=${PATH%:}
	unset RUBY_ROOT RUBY_ENGINE RUBY_VERSION RUBYOPT
	hash -r
}

function chruby_use()
{
	if [[ ! -x "$1/bin/ruby" ]]; then
		echo "chruby: $1/bin/ruby not executable" >&2
		return 1
	fi

	[[ -n "$RUBY_ROOT" ]] && chruby_reset

	export RUBY_ROOT="$1"
	export RUBYOPT="$2"
	export PATH="$RUBY_ROOT/bin:$PATH"

	eval `ruby - <<EOF
require 'rubygems'
puts "export RUBY_ENGINE=#{defined?(RUBY_ENGINE) ? RUBY_ENGINE : 'ruby'};"
puts "export RUBY_VERSION=#{RUBY_VERSION};"
puts "export GEM_ROOT=#{Gem.default_dir.inspect};"
EOF`

	if [[ ! $UID -eq 0 ]]; then
		export GEM_HOME="$HOME/.gem/$RUBY_ENGINE/$RUBY_VERSION"
		export GEM_PATH="$GEM_HOME:$GEM_ROOT"
		export PATH="$GEM_HOME/bin:$GEM_ROOT/bin:$PATH"
	fi
}

function chruby()
{
	local ruby_path

	case "$1" in
		-h|--help)
			echo "usage: chruby [RUBY|VERSION|system] [RUBY_OPTS]"
			;;
		"")
			local star

			for dir in ${RUBIES[@]}; do
				if [[ "$dir" == "$RUBY_ROOT" ]]; then star="*"
				else                                  star=" "
				fi

				echo " $star $(basename "$dir")"
			done
			;;
		system) chruby_reset ;;
		*)
			for dir in ${RUBIES[@]}; do
				if [[ `basename "$dir"` == *$1* ]]; then
					shift
					chruby_use "$dir" "$*"
					return $?
				fi
			done

			echo "chruby: unknown Ruby: $1" >&2
			return 1
			;;
	esac
}
